### ISSUE TITLE
bazel crashes when previous builds event upload is still going on after a remote build

### ISSUE THREAD

**Issue by amarshat** [bazelbuild/bazel#11408]
### Description of the problem / feature request:

We have a corporate RBE setup and we set following flags (among others for a remote build)
```
--remote_executor
--bes_backend
--bes_upload_mode=fully_async
--cpu=k8
```

When we do a build with RBE and immediately after that we do a local build with a flag which can discard the analysis cache (for example `--cpu=darwin`), Bazel crashes with following stack trace.

```
ERROR: bazel crash in async thread:
java.lang.IllegalStateException: RequestMetadata not set in current context.
	at com.google.devtools.build.lib.remote.util.TracingMetadataUtils.fromCurrentContext(TracingMetadataUtils.java:95)
	at com.google.devtools.build.lib.remote.GrpcCacheClient.lambda$findMissingDigests$1(GrpcCacheClient.java:210)
	at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.doFallback(AbstractCatchingFuture.java:175)
	at com.google.common.util.concurrent.AbstractCatchingFuture$AsyncCatchingFuture.doFallback(AbstractCatchingFuture.java:162)
	at com.google.common.util.concurrent.AbstractCatchingFuture.run(AbstractCatchingFuture.java:107)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:711)
	at com.google.common.util.concurrent.AggregateFuture$RunningState.handleException(AggregateFuture.java:178)
	at com.google.common.util.concurrent.AggregateFuture$RunningState.handleOneInputDone(AggregateFuture.java:234)
	at com.google.common.util.concurrent.AggregateFuture$RunningState.access$200(AggregateFuture.java:92)
	at com.google.common.util.concurrent.AggregateFuture$RunningState$1.run(AggregateFuture.java:147)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1024)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:866)
	at com.google.common.util.concurrent.AbstractFuture.setFuture(AbstractFuture.java:752)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleAsyncTask.afterRanInterruptibly(TrustedListenableFutureTask.java:168)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleAsyncTask.afterRanInterruptibly(TrustedListenableFutureTask.java:143)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:80)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
[6 / 9] GoStdlib external/io_bazel_rules_go/darwin_amd64_pure_stripped/stdlib%/pkg; 15s processwrapper-sandbox

Server terminated abruptly (error code: 14, error message: 'Socket closed', log file: '/root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/server/jvm.out')
```

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

- a full remote build with `bes_events_mode` as fully_async
- do a local build

The crash is bound to happen if the events upload from the previous run are still happening.

### What operating system are you running Bazel on?
`macOS`

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

release 3.1.0

###  Have you found anything relevant by searching the web?

Looks related: https://github.com/bazelbuild/bazel/issues/11392

**Comment by meisterT** [comment]
cc @ericfelly

**Comment by amarshat** [comment]
We did some investigation on this, it looks like, if the `remote_retries` is higher (~15), and the remote timeout (`--bes_timeout`) is less than the minimum time it takes for us to upload events, bazel keeps retrying this and crashes. If we increase bes_timeout, the crash can be prevented.

**Comment by ericfelly** [comment]
@michaeledgar do you know how to repro this?

**Comment by brentleyjones** [comment]
This might be fixed in 4.1.0+?

@amarshat, is this still an issue for y'all?

**Comment by coeuvre** [comment]
Closing since this should be fixed.
